darshan-runtime/lib/darshan-core.c

```diff
@@ -470,13 +474,41 @@ void darshan_core_shutdown(int write_log)
     internal_timing_flag = final_core->config.internal_timing_flag;
 
 #ifdef __DARSHAN_ENABLE_MMAP_LOGS
-    /* remove the temporary mmap log files */
-    /* NOTE: this unlink is not immediate as it must wait for the mapping
-     * to no longer be referenced, which in our case happens when the
-     * executable exits. If the application terminates mid-shutdown, then
-     * there will be no mmap files and no final log file.
+    /* Flush memory-mapped data to the underlying file and then duplicate the
+     * mmap log file, in case an interrupt happens before the completion of
+     * this subroutine, leaving the log file corrupted. See github issue #1052.
      */
-    unlink(final_core->mmap_log_name);
+    int sys_page_size = sysconf(_SC_PAGESIZE);
```

Comment:
This code block uses two methods to calculate the size, one based on Darshan parameters and sizes, and one based on the file size (via lseek).  It seems like we could just do one or the other and reuse it for both purposes?  Of the two I would probably vote for just using the file size (lseek end position) so that nothing has to be updated here if the Darshan file format changes.

Reply:
Calling `msync()` before `lseek()` is to ensure all pending writes are flushed to
the mmap file, so we can then safely call `lseek()` to obtain the file size.

To avoid calculation of `mmap_size` in multiple places, a solution can be to save
`mmap_size` as a member of `struct darshan_core_runtime`.

Reply:
So, removing the call to `lseek` and just use `mmap_size` to read and write?

Reply:
I'm advocating for removing the mmap_size calculation.  I would be fine if we *just* used the size reported by lseek (both to determine how much to flush, and to determine how much to copy), or else added a field to the runtime structure to keep up with the size.